proposals/0002.md

```diff
@@ -0,0 +1,272 @@
+method and get its cause.
+
+`source` should return `None` when the error is the root cause and doesn't wrap another error.
+It's also impossible to return a wrapped error when it doesn't implement `Error`.
+
+### The bevavior of the `Display`
+
+The `Display` implementation should briefly describe the error.
+It should never include information provided by the error returned from the `source`.
+Its output should be formatted as a single line, preferably a single sentence.
+If it's a single sentence, it should start with a lowercase letter and end with no interpunction.
+
+Examples:
+
+`failed to load the file '/tmp/file'` **good**
+
+`failed to load the file '/tmp/file': file not found`
+**bad**: includes information provided by its `source`
+
+`Failed to load the file '/tmp/file'!`
+**bad**: a single sentence starting with a capital letter and ending with interpunction
+
+`Failed to load the file. Path: '/tmp/file'.`
+**ugly**: it has two sentences
+
+## Printing a chain of `Error`s
+
+When all the errors in a chain behave correctly, insightful and well formatted messages can be
+easily provided to the user.
+The error can be iterated by recursively calling `source` to print each layer and provide the full
+chain of errors that caused the failure.
+Due to strict `Display` implementation rules there's a lot of creative freedom when it comes to
+chain formatting.
+The `source` traverasal can be done manually, with `anyhow::Chain` or with very similar
+`std::error::Chain` as soon as it's stabilized.
+
+An example of the manual chain printing:
+
+```rust
+println!("An error occured: {}", error);
+let mut parent = &error;
+while let Some(cause) = parent.source() {
+    println!("  caused by: {}", cause);
+    parent = cause;
+}
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file
+  caused by: failed to load the file '~/.app_config'
+  caused by: file not found
+```
+
+An example of the chain printing with `anyhow::Chain`:
+
+```rust
+print!("An error occured: {}", error);
+anyhow::Chain::new(&error).for_each(|e| print!(": {}", e);
+println!();
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file: failed to load the file '~/.app_config': file not found
+```
+
+# Recommendation
+
+These examples of building errors serve as a practical guide.
+There are some common patterns of error structures and all of them can be implemented conveniently
+and correctly using `thiserror` and `anyhow`.
+
+All the examples use `enum`s, but it's fine to use `struct`s instead when there's only one variant.
+Usually it's not very useful, because it reduces flexibility, but doesn't simplify code much.
+
+## Root cause errors
+
+These are errors that don't wrap other errors inside.
+They may contain multiple structures and a lot of details, but their `source` always returns `None`.
+A root error can be just a variant of a bigger error with other variants, some of which may be even
```

Comment:
I don't understand this point. If it's a variant containing a bigger error then isn't the bigger error the cause for this variant in question?

Reply:
Hopefully fixed :)